[SPARK-14321][SQL] Reduce date format cost and string-to-date cost in date functions

rajeshbalamohan opened the pull request:
## What changes were proposed in this pull request?

Here is the generated code snippet when executing date functions. SimpleDateFormat is fairly expensive and can show up bottleneck when processing millions of records. It would be better to instantiate it once.

```
/* 066 */     UTF8String primitive5 = null;
/* 067 */     if (!isNull4) {
/* 068 */       try {
/* 069 */         primitive5 = UTF8String.fromString(new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(
/* 070 */             new java.util.Date(primitive7 * 1000L)));
/* 071 */       } catch (java.lang.Throwable e) {
/* 072 */         isNull4 = true;
/* 073 */       }
/* 074 */     }
```

With modified code, here is the generated code

```
/* 010 */   private java.text.SimpleDateFormat sdf2;
/* 011 */   private UnsafeRow result13;
/* 012 */   private org.apache.spark.sql.catalyst.expressions.codegen.BufferHolder bufferHolder14;
/* 013 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter rowWriter15;
/* 014 */
...
...
/* 065 */     boolean isNull0 = isNull3;
/* 066 */     UTF8String primitive1 = null;
/* 067 */     if (!isNull0) {
/* 068 */       try {
/* 069 */         if (sdf2 == null) {
/* 070 */           sdf2 = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
/* 071 */         }
/* 072 */         primitive1 = UTF8String.fromString(sdf2.format(
/* 073 */             new java.util.Date(primitive4 * 1000L)));
/* 074 */       } catch (java.lang.Throwable e) {
/* 075 */         isNull0 = true;
/* 076 */       }
/* 077 */     }
```

Similarly Calendar.getInstance was used in DateTimeUtils which can be lazily inited.
## How was this patch tested?

org.apache.spark.sql.catalyst.expressions.DateExpressionsSuite,org.apache.spark.sql.catalyst.util.DateTimeUtilsSuite
Also tried with couple of sample SQL queries with single executor (6GB) which showed good improvement with the fix.

rbalamohan committed:
SPARK-14321. [SQL]  Reduce date format cost and string-to-date cost in date functions

AmplabJenkins commented:
Can one of the admins verify this patch?

davies commented:
Jenkins, OK to test

SparkQA commented:
**[Test build #2721 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2721/consoleFull)** for PR 12105 at commit [`6fd07db`](https://github.com/apache/spark/commit/6fd07db11b5c9eed795dde11177f1c245a6fef16).

SparkQA commented:
**[Test build #2721 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2721/consoleFull)** for PR 12105 at commit [`6fd07db`](https://github.com/apache/spark/commit/6fd07db11b5c9eed795dde11177f1c245a6fef16).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
The problem is that this object is not thread-safe. Unless you know this instance is only accessed by one thread, this potentially fails when accessed concurrently.

rajeshbalamohan commented:
SDF declared in the generated code is not shared in multiple threads.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:
```diff
@@ -59,6 +59,8 @@ object DateTimeUtils {
   final val toYearZero = to2001 + 7304850
   final val TimeZoneGMT = TimeZone.getTimeZone("GMT")
 
+  lazy val c = Calendar.getInstance(TimeZone.getTimeZone("GMT"))
```
Yes if SimpleDateFormat isn't used in multiple threads and you're sure it can't be, then caching it in the instance is OK. 

This however isn't thread-safe. The Calendar would be shared across threads.

rbalamohan committed:
SPARK-14321. [SQL]  Reduce date format cost in date functions

rajeshbalamohan commented:
Agreed. Thanks @srowen . Reverted calendar changes in DateTimeUtils in recent commit.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -391,13 +391,14 @@ abstract class UnixTime extends BinaryExpression with ExpectsInputTypes {
             ${ctx.javaType(dataType)} ${ev.value} = ${ctx.defaultValue(dataType)};
           """
         } else {
+          val formatter = ctx.freshName("sdf")
```
Pardon my ignorance but is "sdf" the right name here? I admit I am not familiar with how the code gen works, but just looking at how it seems to operate, formatter seems to be a different reference than sdf.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -416,11 +419,17 @@ abstract class UnixTime extends BinaryExpression with ExpectsInputTypes {
           s"""
             try {
+              if ($formatter == null ||
+                  !$formatter.toPattern().equals("$format.toString()")) {
```
So, this raises an interesting point ... the SimpleDateFormat you're caching is a reference used across logically different code branches that need different formats. Are we sure this is a win? because every time a different format is needed you have to switch anyway. I might just be paranoid

davies replied:
This optimization seems useless to me too, I'd like to not have it.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -519,7 +530,10 @@ case class FromUnixTime(sec: Expression, format: Expression)
           if (!${ev.isNull}) {
             try {
-              ${ev.value} = UTF8String.fromString(new $sdf("${constFormat.toString}").format(
+              if ($sdfTerm == null) {
```
Does this not need the same guard above to check the format? I suspect I might misunderstand the logic. CC @davies for a look too

davies replied:
format is an constant here.

davies commented:
@rajeshbalamohan Could you also update the eval() of UnixTime, FromUnixTime and DateFormatClass?

srowen commented:
Ping @rajeshbalamohan

srowen commented:
@rajeshbalamohan are you still working on this?

rajeshbalamohan commented:
Sorry about the delay. I will share the update patch today

rbalamohan committed:
[SPARK-14321][SQL]  Reduce date format cost in date functions

rbalamohan committed:
[SPARK-14321][SQL]  Reduce date format cost in date functions

rajeshbalamohan commented:
Revised the patch addressing comments. Fixed eval() of UnixTime, FromUnixTime.  Haven't changed eval in DateFormatClass as i am not sure if format can change in between.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -355,6 +355,7 @@ abstract class UnixTime extends BinaryExpression with ExpectsInputTypes {
   override def nullable: Boolean = true
 
   private lazy val constFormat: UTF8String = right.eval().asInstanceOf[UTF8String]
+  private var formatter: SimpleDateFormat = _
```
Not sure this is worth changing, but `...: DateFormat = null`? I actually am not familiar with the convention for when to use _ vs null. I thought = _ was mostly for defs.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -368,7 +369,10 @@ abstract class UnixTime extends BinaryExpression with ExpectsInputTypes {
           if (constFormat != null) {
-            Try(new SimpleDateFormat(constFormat.toString).parse(
+            if (formatter == null) {
+              formatter = Try(new SimpleDateFormat(constFormat.toString)).getOrElse(null)
```
`.orNull` is available? but, this will just cause an NPE in the next line, which will also get squashed. Do we really want to hide both of these? why not let the error fly?

rajeshbalamohan replied:
Didn't want to throw the error back as it would break the earlier functionality.  Eearlier it was returning null when any exception (i.e, could be constFormat being null, or parsing error) was thrown.  

Creating the formatter upfront in the recent commit and handling null earlier itself, to have minimal changes.

rbalamohan committed:
[SPARK-14321][SQL]  Reduce date format cost in date functions

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -391,21 +395,24 @@ abstract class UnixTime extends BinaryExpression with ExpectsInputTypes {
               try {
-                $sdf $formatter = new $sdf("$fString");
+                if ($formatter == null) {
+                  $formatter = new $sdf("$fString");
```
Hm, but doesn't this generated code then need the same guards as above? I think I may not know enough to review this

rbalamohan committed:
[SPARK-14321][SQL]  Reduce date format cost in date functions

rajeshbalamohan commented:
In the generated code, it returns null if constFormat == null.  So it is not required to change the generated code.

HyukjinKwon commented:
This is trivial but would be great if the title follows `[SPARK-XXXXX][SQL]` format as described in https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:
```diff
@@ -391,21 +393,24 @@ abstract class UnixTime extends BinaryExpression with ExpectsInputTypes {
           """
         } else {
+          val formatter = ctx.freshName("formatter")
+          ctx.addMutableState(sdf, formatter, s"""$formatter = null;""")
```
is it possible to create the formatter here?

rajeshbalamohan replied:
yes. Creating the formatter here did not create any issues.

cloud-fan commented:
hi @rajeshbalamohan , sorry for the delay,  do you wanna try this again?

sameeragarwal commented:
@rajeshbalamohan thanks again for working on this; it'd be great to have this in 2.0. Do you have cycles to bring to this up to date and address reviewer comments?

rajeshbalamohan commented:
Sorry about the delay in responding to this. Will try to rebase and post the patch asap.

zzcclp commented:
@rajeshbalamohan , Is there any progress yet? We meet the same performance issues when formatting date or operating string-to-date .

rajeshbalamohan commented:
Patch went stale for master branch and got little messy in my system. I have created https://github.com/apache/spark/pull/13522 which is rebased to master. Will close this after view.

srowen commented:
Can you close this PR then?
